lib/charms/tls_certificates_interface/v2/tls_certificates.py

```diff
@@ -1159,6 +1168,84 @@ def _revoke_certificates_for_which_no_csr_exists(self, relation_id: int) -> None
+        """Filters the requirer's units csrs.
+
+        Keeps the ones for which no certificate was provided.
+
+        Returns:
+            list: List of dictonries that contain the unit's csrs
+            that don't have a certificate issued.
+        """
+        all_unit_csr_mappings = copy.deepcopy(self.get_requirer_csrs_by_unit())
+        for unit_csrs in all_unit_csr_mappings:
+            base64_keys_list = list(
+                map(
+                    lambda item: base64.b64encode(item.encode()).decode(),
+                    self.get_issued_certificates(unit_csrs["relation_id"])[  # type: ignore[arg-type]
+                        unit_csrs["application_name"]  # type: ignore[index]
+                    ].keys(),
+                )
+            )
+            for csr in unit_csrs["unit_csrs"]:  # type: ignore[union-attr]
+                if csr["certificate_signing_request"] in base64_keys_list:  # type: ignore[index]
+                    unit_csrs["unit_csrs"].remove(csr)  # type: ignore[union-attr, arg-type]
+            if len(unit_csrs["unit_csrs"]) == 0:  # type: ignore[arg-type]
+                all_unit_csr_mappings.remove(unit_csrs)
+        return all_unit_csr_mappings
+
+    def get_requirer_csrs_by_unit(
+        self, relation_id: Optional[int] = None
+    ) -> List[Dict[str, Union[int, str, List[Dict[str, str]]]]]:
+        """Returns a list of requirers' CSRs grouped by unit.
+
+        It returns CSRs from all relations if relation_id is not specified.
+        CSRs are returned per relation id, application name and unit name.
+        CSRs are returned in base64 encoded format.
+
+        Returns:
+            list: List of dictonries that contain the unit's csrs
+            with the following information
+            relation_id, application_name and unit_name.
+        """
+        unit_csr_mappings: List[Dict[str, Union[int, str, List[Dict[str, str]]]]] = []
+
+        relations = (
+            [
+                relation
+                for relation in self.model.relations[self.relationship_name]
+                if relation.id == relation_id
+            ]  # noqa: E501
+            if relation_id is not None
+            else self.model.relations.get(self.relationship_name, [])
+        )
+
+        for relation in relations:
+            for unit in relation.units:
+                requirer_relation_data = _load_relation_data(relation.data[unit])
+                unit_csrs_list = requirer_relation_data.get("certificate_signing_requests", [])
+                transformed_unit_csrs_list = list(
+                    map(
+                        lambda item: {
+                            "certificate_signing_request": base64.b64encode(
+                                item["certificate_signing_request"].encode()
+                            ).decode("utf-8")
+                        },
+                        unit_csrs_list,
+                    )
+                )
```

Comment:
This block is quite complex. All those data manipulations make it quite hard to understand. If we could at least centralise the necessary decode/encode in 1 method and properly explain what it is we are doing it would help a bit.

Reply:
As above, this is pushed down to the operator.